### ISSUE TITLE
kernel: add @aliasTo decorator

### ISSUE THREAD

**Issue by xenoterracide** [aurelia/aurelia#862]
we have quite a few aliases, on things that should be able to jitRegister, it's frustrating to have to go through and also register them by hand suggestion.

```ts
@aliasTo(MyKey)
export class Foo {...{

container.get(Foo); // works
container.get(MyKey); //works
```

```ts
@transient()
@aliasTo(MyKey)
export class Foo {...{

container.get(Foo); // works, new one every time
container.get(MyKey); //works, new one every time
```
and our real use case
```ts
export const Handler = DI.createInterface<Handler>().noDefault();

interface Handler {}

@aliasTo(Handler)
export class Foo {}
@aliasTo(Handler)
export class Bar {}
@aliasTo(Handler)
export class Baz {}

container.get(Foo); // works
container.getAll(Handler); //works
```

**Comment by thomas-darling** [comment]
Let's say we have a class, and that class needs e.g. the router injected.
As I understand it, we currently have to apply the `@IRouter` decorator to that parameter - which I assume has something to do with limitations around how TypeScript emits metadata for interface types. Correct?

```ts
@inject()
export class MyCustomElement {
    constructor(@IRouter router: IRouter)
    {
      // ...
    }
}

```

Now, I fully understand the fundamental desire to use interfaces as DI keys - it makes perfect sense, as it allows the implementation to be completely substituded.

However, in practice, I very rarely need this flexibility when building an app - what I do need, however, is to inject the router, or some other framework type, into _lots_ of different components or services. And having to specify those decorators on all those constructor parameters - it just feels really messy and redundant.

Therefore, I would _really_ like to be able to just specify the parameter type as e.g. `Router`, like we do in v1, and then get the same instance injected, as if the type was specified as `IRouter`.

Am I correct that this issue, once resolved, would allow me to do just that?
And if not, is there some other way to acheive this, such that e.g. `IRouter` and `Router` both resolve to the same instance?

**Comment by zewa666** [comment]
the purpose for the Interface, coming from the testing perspective is clear, but if you need a shorthand how about registering an alias on any of your preferred keys?

```
Registration.alias(originalKey: any, aliasKey: any): IRegistration
```

https://docs.aurelia.io/getting-to-know-aurelia/dependency-injection-di

**Comment by thomas-darling** [comment]
Oh, really not sure how I missed that `Registration.alias` method - Thanks! 🙂 

Regarding the testing use case:

I fully agree that using an interface is, at least from an academic point of view, the most correct approach. However, in practice, I could just as well register a `MockRouter` with `Router` as the key, and it would achieve the exact same thing. 

Interfaces only really provide value, in a scenario where I want to completely exclude the code for the default implementation from the build, and substitude my own implementation instead - or as in the example here, have multiple implementations for the same key, at the same time. And those are absolutely valid scenarios - just not very common ones, relatively speaking.

Also, another approach, which I believe would achieve the same thing, but without the drawbacks associated with interfaces, would be to use an abstract class instead of an interface - I believe those are emitted in metadata, just like normal classes.

But anyways... back to the issue at hand.
Maybe I'm missing something here, but doesn't this example imply, that the the behavior I want, would work by default too?
```ts
container.get(Foo); // works
container.getAll(Handler); //works
```
I mean, one of the instances returned by `container.getAll(Handler)` would be the same instance returned by `container.get(Foo)`, right? And if that is true, I'd assume `container.get(Handler)` and `container.get(Foo)` would return the same instance too - assuming of course, that only a single implementation has been registered.

**Comment by zewa666** [comment]
the difference with interfaces is that exactly no logic is hidden but only a formal contract is declared. which is a huge thing ones things start to grow. how do you know what MockRouter is doing with static decos etc when you use it as key? or if anyone modified and added hidden logic? Interfaces, especially in ts cant really do anything as at the end they are just naked Symbols. yes its academic but there is a good reason behind that ;)

as to your last point, I must confess I dont get what you meant :) the point of aliasing was to have an additional key pointing to another already existing key and thus having the same resolver. This, in order to save your `@IRouter` import and argument decorator.

**Comment by thomas-darling** [comment]
Well, there are both pros and cons to either approach 🙂 

Regarding my last point:

Maybe I'm just misunderstanding the purpose of the proposed `@aliasTo` decorator...
But my point is, if an `@aliasTo(IRouter)` decorator is applied to the `Router` class, then based on the examples, I'd assume having a constructor parameter of type `Router`, instead of `IRouter`, would also just work, without having to manually add an alias - as I assume constructor injection ultimately ends up calling `container.get`.

But maybe this `@aliasTo` decorator isn't intended to be applied to classes like `Router` at all?
I just thought this was intended to be a general mechanism for associating an implementation with its DI key.

**Comment by zewa666** [comment]
oh lol I didnt realize we're in an issue related to a ln alias deco 🤣 until I now scrolled up. I thought this is the one I saw you posting today about router v1 features in v2.

No the deco discussed here wouldnt help as you'd have to apply it on the router class, which is inside an external package. so the registration alias api would as initially mentioned solve your need I think.

sry for deviating from this issues original topic
